lib/usd/translators/shading/shadingAsset.cpp

```diff
@@ -0,0 +1,383 @@
+//
+// Copyright 2023 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "shadingAsset.h"
+
+#include "shadingTokens.h"
+
+#include <mayaUsd/fileio/utils/readUtil.h>
+#include <mayaUsd/utils/utilFileSystem.h>
+
+#include <pxr/base/arch/hash.h>
+#include <pxr/usd/ar/asset.h>
+#include <pxr/usd/ar/packageUtils.h>
+#include <pxr/usd/ar/resolver.h>
+#include <pxr/usd/sdf/assetPath.h>
+#include <pxr/usd/sdf/layerUtils.h>
+#include <pxr/usd/usd/resolver.h>
+#include <pxr/usd/usdShade/input.h>
+#include <pxr/usd/usdShade/shader.h>
+
+#include <ghc/filesystem.hpp>
+
+#include <sstream>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+static void
+updateAssetPath(std::string assetPath, std::string resolvedPath, SdfAssetPath* resolvedAssetPath)
+{
+    assetPath = TfStringReplace(assetPath, "\\", "/");
+
+    if (resolvedPath.empty())
+        resolvedPath = assetPath;
+    else
+        resolvedPath = TfStringReplace(resolvedPath, "\\", "/");
+
+    *resolvedAssetPath = SdfAssetPath(assetPath, resolvedPath);
+}
+
+static SdfAssetPath handleShaderInput(const UsdShadeInput& usdInput)
+{
+    VtValue val;
+    if (!usdInput.Get(&val))
+        return {};
+
+    if (!val.IsHolding<SdfAssetPath>())
+        return {};
+
+    return val.UncheckedGet<SdfAssetPath>();
+}
+
+static bool handleMissingResolvedPath(SdfAssetPath* resolvedAssetPath)
```

Comment:
This all code taken from usdUVTextureReader,cpp and refactored into multiple functions. You would need to diff with that file by hand to see what changed.